pkg/providers/instancetype/types.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/operator/options"
+)
+
+var (
+	instanceTypeScheme = regexp.MustCompile(`^ecs\.([a-z]+)(\-[0-9]+tb)?([0-9]+).*`)
+)
+
+const (
+	MemoryAvailable = "memory.available"
+	NodeFSAvailable = "nodefs.available"
+)
+
+func NewInstanceType(ctx context.Context, info *ecsclient.DescribeInstanceTypesResponseBodyInstanceTypesInstanceType, kc *v1alpha1.KubeletConfiguration, region string, offerings cloudprovider.Offerings) *cloudprovider.InstanceType {
+
+	it := &cloudprovider.InstanceType{
+		Name:         *info.InstanceTypeId,
+		Requirements: computeRequirements(info, offerings, region),
+		Offerings:    offerings,
+		Capacity:     computeCapacity(ctx, info, kc.MaxPods, kc.PodsPerCore),
+		Overhead: &cloudprovider.InstanceTypeOverhead{
+			KubeReserved:      kubeReservedResources(cpu(info), pods(ctx, info, kc.MaxPods, kc.PodsPerCore), kc.KubeReserved),
+			SystemReserved:    systemReservedResources(kc.SystemReserved),
+			EvictionThreshold: evictionThreshold(memory(ctx, info), ephemeralStorage(info), kc.EvictionHard, kc.EvictionSoft),
+		},
+	}
+	if it.Requirements.Compatible(scheduling.NewRequirements(scheduling.NewRequirement(corev1.LabelOSStable, corev1.NodeSelectorOpIn, string(corev1.Windows)))) == nil {
+		it.Capacity[v1alpha1.ResourcePrivateIPv4Address] = *privateIPv4Address(info)
+	}
+	return it
+}
+
+//nolint:gocyclo
+func computeRequirements(info *ecsclient.DescribeInstanceTypesResponseBodyInstanceTypesInstanceType, offerings cloudprovider.Offerings, region string) scheduling.Requirements {
+	requirements := scheduling.NewRequirements(
+		// Well Known Upstream
+		scheduling.NewRequirement(corev1.LabelInstanceTypeStable, corev1.NodeSelectorOpIn, *info.InstanceTypeId),
+		scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, *info.CpuArchitecture),
+		scheduling.NewRequirement(corev1.LabelOSStable, corev1.NodeSelectorOpIn, string(corev1.Linux)),
+		scheduling.NewRequirement(corev1.LabelTopologyZone, corev1.NodeSelectorOpIn, lo.Map(offerings.Available(), func(o cloudprovider.Offering, _ int) string {
+			return o.Requirements.Get(corev1.LabelTopologyZone).Any()
+		})...),
+		scheduling.NewRequirement(corev1.LabelTopologyRegion, corev1.NodeSelectorOpIn, region),
+		scheduling.NewRequirement(corev1.LabelWindowsBuild, corev1.NodeSelectorOpDoesNotExist),
+		// Well Known to Karpenter
+		scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, lo.Map(offerings.Available(), func(o cloudprovider.Offering, _ int) string {
+			return o.Requirements.Get(karpv1.CapacityTypeLabelKey).Any()
+		})...),
+		// Well Known to AliCloud
+		scheduling.NewRequirement(v1alpha1.LabelInstanceCPU, corev1.NodeSelectorOpIn, fmt.Sprint(*info.CpuCoreCount)),
+		scheduling.NewRequirement(v1alpha1.LabelInstanceCPUManufacturer, corev1.NodeSelectorOpDoesNotExist),
+		scheduling.NewRequirement(v1alpha1.LabelInstanceMemory, corev1.NodeSelectorOpIn, fmt.Sprint(*info.MemorySize)),
```

Comment:
AlibabaCloud doesn't looks like aws, their label is quite simple:
```
  annotations:
    alibabacloud.com/attach-time: "2024-10-10 13:55:31,6"
    alibabacloud.com/create-time: "2024-10-10 13:55:40,16"
    alibabacloud.com/data-disk-serial-id: ""
    csi.volume.kubernetes.io/nodeid: '{"diskplugin.csi.alibabacloud.com":"i-8vbh9l3aj483p8z6db8a","nasplugin.csi.alibabacloud.com":"cn-zhangjiakou.172.26.12.238","ossplugin.csi.alibabacloud.com":"cn-zhangjiakou.172.26.12.238"}'
    k8s.aliyun.com/max-available-ip: "20"
    kubeadm.alpha.kubernetes.io/cri-socket: unix:///var/run/containerd/containerd.sock
    node.alpha.kubernetes.io/ttl: "0"
    node.csi.alibabacloud.com/allocatable-disk: "16"
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: "2024-10-10T05:55:35Z"
  labels:
    ack.aliyun.com: c3c4c4fdc6bf44d889b45c872b0fb7cc4
    alibabacloud.com/ecs-instance-id: i-8vbh9l3aj483p8z6db8a
    alibabacloud.com/nodepool-id: npd690349493e243ff8dd3c98a3106d8fd
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: ecs.g6.xlarge
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: cn-zhangjiakou
    failure-domain.beta.kubernetes.io/zone: cn-zhangjiakou-c
    goatscaler.io/managed: "true"
    goatscaler.io/provision-task-id: asa-8vb0kstbsmm3faefj35d
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: cn-zhangjiakou.172.26.12.238
    kubernetes.io/os: linux
    node.csi.alibabacloud.com/disktype.cloud_auto: available
    node.csi.alibabacloud.com/disktype.cloud_efficiency: available
    node.csi.alibabacloud.com/disktype.cloud_essd: available
    node.csi.alibabacloud.com/disktype.cloud_ssd: available
    node.kubernetes.io/instance-type: ecs.g6.xlarge
    sk: terway
    topology.diskplugin.csi.alibabacloud.com/zone: cn-zhangjiakou-c
    topology.kubernetes.io/region: cn-zhangjiakou
    topology.kubernetes.io/zone: cn-zhangjiakou-c
  name: cn-zhangjiakou.172.26.12.238
  resourceVersion: "32368"
  uid: c3513ad4-b28c-4d8f-a786-c7f3711f7162
```

We should refactor this part based on the real node yaml

Reply:
BTW, we should set the label referring to our requirements in the future

Reply:
These labels are enhanced features. I feel that even the native end does not have.

Reply:
@daimaxiaxie is right,
```bash
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      alpha.kubernetes.io/provided-node-ip: 10.0.13.39
      csi.volume.kubernetes.io/nodeid: '{"ebs.csi.aws.com":"i-0dd47c70a3ef093ef"}'
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: "2024-10-15T06:13:30Z"
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/instance-type: t3.xlarge
      beta.kubernetes.io/os: linux
      eks.amazonaws.com/capacityType: ON_DEMAND
      eks.amazonaws.com/nodegroup: od-managed-node-group-20241015061232984800000015
      eks.amazonaws.com/nodegroup-image: ami-04f969acb2711f0e8
      eks.amazonaws.com/sourceLaunchTemplateId: lt-03224a8ce978cb8a2
      eks.amazonaws.com/sourceLaunchTemplateVersion: "1"
      failure-domain.beta.kubernetes.io/region: us-east-2
      failure-domain.beta.kubernetes.io/zone: us-east-2a
      k8s.io/cloud-provider-aws: 7a74acd98dd0d9576ef2ba721e8750e9
      kubernetes.io/arch: amd64
      kubernetes.io/hostname: ip-10-0-13-39.us-east-2.compute.internal
      kubernetes.io/os: linux
      node.kubernetes.io/instance-type: t3.xlarge
      topology.ebs.csi.aws.com/zone: us-east-2a
      topology.k8s.aws/zone-id: use2-az1
      topology.kubernetes.io/region: us-east-2
      topology.kubernetes.io/zone: us-east-2a
    name: ip-10-0-13-39.us-east-2.compute.internal
    resourceVersion: "43693"
    uid: 9c140d52-a5e7-4cf0-833c-14cbe67a4217
  spec:

```